File: src/main/scala/li/cil/oc/common/inventory/ComponentInventory.scala
package li.cil.oc.common.inventory

import li.cil.oc.OpenComputers
import li.cil.oc.api
import li.cil.oc.api.Driver
import li.cil.oc.api.driver.{DriverItem => ItemDriver}
import li.cil.oc.api.network
import li.cil.oc.api.network.EnvironmentHost
import li.cil.oc.api.network.ManagedEnvironment
import li.cil.oc.api.network.Node
import li.cil.oc.api.util.Lifecycle
import li.cil.oc.integration.opencomputers.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound

import scala.collection.convert.WrapAsScala._
import scala.collection.mutable

trait ComponentInventory extends Inventory with network.Environment {
  lazy val components = Array.fill[Option[ManagedEnvironment]](getSizeInventory)(None)
  protected val updatingComponents = mutable.ArrayBuffer.empty[ManagedEnvironment]

  // ----------------------------------------------------------------------- //

  def host: EnvironmentHost

  // ----------------------------------------------------------------------- //

  def updateComponents() {
    if (updatingComponents.nonEmpty) {
      var i = 0
      // ArrayBuffer.foreach caches the size for performance reasons, but that
      // will cause issues if the list changed during iteration (e.g. because
      // a component removed itself / another component, such as the self-
      // destruct card from Computronics). Also, this list will generally be
      // quite short, so it won't have any noticeable impact, anyway.
      while (i < updatingComponents.size) {
        updatingComponents(i).update()
        i += 1
      }
    }
  }

  // ----------------------------------------------------------------------- //

  def connectComponents() {
    for (slot <- 0 until getSizeInventory if slot >= 0 && slot < components.length) {
      val stack = getStackInSlot(slot)
      if (!stack.isEmpty && components(slot).isEmpty && isComponentSlot(slot, stack)) {
        components(slot) = Option(Driver.driverFor(stack)) match {
          case Some(driver) =>
            Option(driver.createEnvironment(stack, host)) match {
              case Some(component) =>
                applyLifecycleState(component, Lifecycle.LifecycleState.Constructing)
                try {
                  component.load(dataTag(driver, stack))
                }
                catch {
                  case e: Throwable => OpenComputers.log.warn(s"An item component of type '${component.getClass.getName}' (provided by driver '${driver.getClass.getName}') threw an error while loading.", e)
                }
                if (component.canUpdate) {
                  assert(!updatingComponents.contains(component))
                  updatingComponents += component
                }
                Some(component)
              case _ => None
            }
          case _ => None
        }
      }
    }
    // Make sure our node is connected.
    api.Network.joinNewNetwork(node)
    components collect {
      case Some(component) =>
        applyLifecycleState(component, Lifecycle.LifecycleState.Initializing)
        connectItemNode(component.node)
        applyLifecycleState(component, Lifecycle.LifecycleState.Initialized)
    }
  }

  def disconnectComponents() {
    components collect {
      case Some(component) =>
        applyLifecycleState(component, Lifecycle.LifecycleState.Disposing)
        if (component.node != null) component.node.remove()
        applyLifecycleState(component, Lifecycle.LifecycleState.Disposed)
    }
  }

  // ----------------------------------------------------------------------- //

  override def save(nbt: NBTTagCompound) = {
    saveComponents()
    super.save(nbt) // Save items after updating their tags.
  }

  def saveComponents() {
    for (slot <- 0 until getSizeInventory) {
      val stack = getStackInSlot(slot)
      if (!stack.isEmpty) {
        components(slot) match {
          case Some(component) =>
            // We're guaranteed to have a driver for entries.
            save(component, Driver.driverFor(stack), stack)
          case _ => // Nothing special to save.
        }
      }
    }
  }

  // ----------------------------------------------------------------------- //

  override def getInventoryStackLimit = 1

  override protected def onItemAdded(slot: Int, stack: ItemStack) = if (isComponentSlot(slot, stack)) {
    Option(Driver.driverFor(stack)).foreach(driver =>
      Option(driver.createEnvironment(stack, host)) match {
        case Some(component) => this.synchronized {
          components(slot) = Some(component)
          applyLifecycleState(component, Lifecycle.LifecycleState.Constructing)
          try {
            component.load(dataTag(driver, stack))
          } catch {
            case e: Throwable => OpenComputers.log.warn(s"An item component of type '${component.getClass.getName}' (provided by driver '${driver.getClass.getName}') threw an error while loading.", e)
          }
          if (component.canUpdate) {
            assert(!updatingComponents.contains(component))
            updatingComponents += component
          }
          applyLifecycleState(component, Lifecycle.LifecycleState.Initializing)
          connectItemNode(component.node)
          applyLifecycleState(component, Lifecycle.LifecycleState.Initialized)
          save(component, driver, stack)
        }
        case _ => // No environment (e.g. RAM).
      })
  }

  override protected def onItemRemoved(slot: Int, stack: ItemStack) {
    // Uninstall component previously in that slot.
    components(slot) match {
      case Some(component) => this.synchronized {
        // Note to self: we have to remove the node from the network *before*
        // saving, to allow file systems to close their handles before they
        // are saved (otherwise hard drives would restore all handles after
        // being installed into a different computer, even!)
        components(slot) = None
        updatingComponents -= component
        applyLifecycleState(component, Lifecycle.LifecycleState.Disposing)
        Option(component.node).foreach(_.remove())
        Option(Driver.driverFor(stack)).foreach(save(component, _, stack))
        // However, nodes then may add themselves to a network again, to
        // ensure they have an address that gets sent to the client, used
        // for associating some components with each other. So we do it again.
        // TODO Should be possible to avoid this with lifecycle state now.
        Option(component.node).foreach(_.remove())
        applyLifecycleState(component, Lifecycle.LifecycleState.Disposed)
      }
      case _ => // Nothing to do.
    }
  }

  def isComponentSlot(slot: Int, stack: ItemStack) = true

  protected def connectItemNode(node: Node) {
    if (this.node != null && node != null) {
      this.node.connect(node)
    }
  }

  protected def dataTag(driver: ItemDriver, stack: ItemStack) =
    Option(driver.dataTag(stack)).getOrElse(Item.dataTag(stack))

  protected def save(component: ManagedEnvironment, driver: ItemDriver, stack: ItemStack): Unit = {
    try {
      val tag = dataTag(driver, stack)
      // Clear the tag compound before saving to get the same behavior as
      // in tile entities (otherwise entries have to be cleared manually).
      for (key <- tag.getKeySet.map(_.asInstanceOf[String])) {
        tag.removeTag(key)
      }
      component.save(tag)
    } catch {
      case e: Throwable => OpenComputers.log.warn(s"An item component of type '${component.getClass.getName}' (provided by driver '${driver.getClass.getName}') threw an error while saving.", e)
    }
  }

  protected def applyLifecycleState(component: AnyRef, state: Lifecycle.LifecycleState): Unit = component match {
    case lifecycle: Lifecycle => lifecycle.onLifecycleStateChange(state)
    case _ =>
  }
}

File: src/main/scala/li/cil/oc/common/inventory/ComponentInventory.java
package li.cil.oc.common.inventory;

import li.cil.oc.OpenComputers;
import li.cil.oc.api.Driver;
import li.cil.oc.api.driver.DriverItem;
import li.cil.oc.api.network.Environment;
import li.cil.oc.api.network.EnvironmentHost;
import li.cil.oc.api.network.EnvironmentItem;
import li.cil.oc.api.network.Node;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.ITickable;
import net.minecraftforge.common.util.INBTSerializable;
import net.minecraftforge.items.IItemHandler;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

public final class ComponentInventory implements ItemHandlerHosted.ItemHandlerHost, INBTSerializable<NBTTagList> {
    public interface ComponentInventoryHost extends EnvironmentHost {
        Node getItemNode();

        IItemHandler getComponentItems();
    }

    // ----------------------------------------------------------------------- //

    private final ComponentInventoryHost host;
    private final EnvironmentItem[] environments;
    private final List<ITickable> updatingEnvironments = new ArrayList<>();

    // ----------------------------------------------------------------------- //

    public ComponentInventory(final ComponentInventoryHost host) {
        this.host = host;
        environments = new EnvironmentItem[host.getComponentItems().getSlots()];
    }

    @Nullable
    public Environment getEnvironment(final int slot) {
        return environments[slot];
    }

    public void update() {
        // No foreach to allow components to remove themselves in their update.
        for (int i = updatingEnvironments.size() - 1; i >= 0; i--) {
            if (i < updatingEnvironments.size()) { // Just in case an update triggered a multi-remove.
                updatingEnvironments.get(i).update();
            }
        }
    }

    public void initialize() {
        dispose();

        final IItemHandler inventory = host.getComponentItems();
        for (int slot = 0; slot < inventory.getSlots(); slot++) {
            final ItemStack stack = inventory.getStackInSlot(slot);
            if (stack.isEmpty()) {
                continue;
            }

            final DriverItem driver = Driver.driverFor(stack, host.getClass());
            if (driver == null) {
                continue;
            }

            final EnvironmentItem environment = driver.createEnvironment(stack, host);
            if (environment == null) {
                continue;
            }

            environments[slot] = environment;

            if (environment instanceof ITickable) {
                updatingEnvironments.add((ITickable) environment);
            }
        }
    }

    public void dispose() {
        for (int slot = 0; slot < environments.length; slot++) {
            final EnvironmentItem environment = environments[slot];
            if (environment == null) {
                continue;
            }

            environments[slot] = null;

            if (environment instanceof ITickable) {
                updatingEnvironments.remove(environment);
            }

            final Node node = environment.getNode();
            if (node != null) {
                node.remove();
            }

            environment.onDispose();
        }
    }

    public void connect() {
        for (final EnvironmentItem environment : environments) {
            if (environment == null) {
                continue;
            }

            final Node node = environment.getNode();
            if (node == null) {
                continue;
            }

            host.getItemNode().connect(environment.getNode());
        }
    }

    @Override
    public void onItemAdded(final int slot, final ItemStack stack) {
        assert environments[slot] == null : "adding item without having removed previous one";
        onItemRemoved(slot, ItemStack.EMPTY.copy()); // Handle assert not holding as gracefully as possible.

        final DriverItem driver = Driver.driverFor(stack, host.getClass());
        if (driver == null) {
            return;
        }

        final EnvironmentItem environment = driver.createEnvironment(stack, host);
        if (environment == null) {
            return;
        }

        environments[slot] = environment;

        if (environment instanceof ITickable) {
            updatingEnvironments.add((ITickable) environment);
        }

        environment.onInstalled(stack);
    }

    @Override
    public void onItemRemoved(final int slot, final ItemStack stack) {
        final EnvironmentItem environment = environments[slot];
        if (environment != null) {
            if (environment instanceof ITickable) {
                updatingEnvironments.remove(environment);
            }

            environments[slot] = null;

            final Node node = environment.getNode();
            if (node != null) {
                node.remove();
            }

            environment.onUninstalled(stack);

            environment.onDispose();
        }
    }

    // ----------------------------------------------------------------------- //
    // INBTSerializable

    @Override
    public NBTTagList serializeNBT() {
        final NBTTagList nbt = new NBTTagList();
        for (final EnvironmentItem environment : environments) {
            if (environment != null) {
                nbt.appendTag(environment.serializeNBT());
            } else {
                nbt.appendTag(new NBTTagCompound());
            }
        }
        return nbt;
    }

    @Override
    public void deserializeNBT(final NBTTagList nbt) {
        initialize();

        if (nbt.tagCount() != environments.length) {
            OpenComputers.log().warn("component count mismatch; not loading component states.");
            return;
        }

        for (int slot = 0; slot < nbt.tagCount(); slot++) {
            if (environments[slot] != null) {
                environments[slot].deserializeNBT(nbt.getCompoundTagAt(slot));
            }
        }
    }
}
